Add config to allow permanent session in cookie

colinrotherham opened the pull request:
I'm currently working across two teams with long form journeys.

### A quick prototyping scenario

1. Enter lots and lots of form data
1. Reach the _Nth_ page in a journey
1. Edit a view or route

### What happens?

1. The prototype is restarted by nodemon
1. All typed in session data is now lost 😭 

### What's in this pull request

I've added the following config:

```js
{
  // Enable cookie-based session store (4KB limit, persists on restart)
  useCookieSessionStore: 'false'
}
```

When set to `true` it swaps _express-session_ with  Mozilla _[client-sessions](https://www.npmjs.com/package/client-sessions)_ so retains compatibility with `req.session` and persists all session data in a cookie.

colinrotherham commented:
Rebased with latest from `master`

colinrotherham commented:
Renamed the option to `useCookieSessionStore`

colinrotherham commented:
As discussed on https://github.com/alphagov/govuk-prototype-kit/issues/472 there were some security concerns around file sessions stored on the server.

e.g. if Heroku was hacked, your encrypted sessions can be stolen and brute forced.

With this approach we avoid that:

1. Data is persisted when the prototype restarts
2. Data is persisted if the prototype is deleted and cloned again
3. Data is encrypted by Mozilla **[client-sessions](https://www.npmjs.com/package/client-sessions)**
4. Data isn't stored on Heroku or on disk, it's in a cookie

nickcolley commented:
I have tested this by:

1. Going through the [pass data example](http://localhost:3000/docs/examples/pass-data)
2. Confirming the session has my inputted data
3. Restarting the server
4. Session is lost

1. Turning on 'useCookieSessionStore'
2. Going through previous 1-3
3. Session persists

I have also tested the 'clear data' function in both modes, and confirmed the session was removed.

fofr commented:
This will be really useful for us at DfE 👍

colinrotherham commented on server.js:
```diff
@@ -182,23 +183,36 @@ app.use(function (req, res, next) {
+// Session uses service name to avoid clashes with other prototypes
+const sessionName = 'govuk-prototype-kit-' + (Buffer.from(config.serviceName, 'utf8')).toString('hex')
+let sessionOptions = {
+  secret: config.serviceName,
```
@hannalaakso Yep, no problem.

Had to remove the randomisation, otherwise you're locked out of your session on restart 😳

colinrotherham commented on app/config.js:
```diff
@@ -15,6 +15,9 @@ module.exports = {
   // Automatically stores form data, and send to all views
   useAutoStoreData: 'true',
 
+  // Enable cookie-based session store (4KB limit, persists on restart)
```
@hannalaakso Yeah I'll add a comment.

Yup that's it. Node.js will happily send a cookie that's too big, but the browser will ignore it and keep the last cookie it had that was under the 4KB limit for the domain.

colinrotherham commented:
Hi @hannalaakso, thanks for the review.

I've addressed both your comments and rebased from `master`.

In future (since the session is mainly text) a little _gzip_ magic could be used to compress it down. It'll reduce cookie size down to about 65% of the original.

joelanman commented on server.js:
```diff
@@ -182,23 +183,36 @@ app.use(function (req, res, next) {
+// Session uses service name to avoid clashes with other prototypes
+const sessionName = 'govuk-prototype-kit-' + (Buffer.from(config.serviceName, 'utf8')).toString('hex')
+let sessionOptions = {
+  secret: sessionName,
```
not sure if there's an advantage to using a predictable string for the 'secret' of the cookie? I think it should be somewhat secret?

joelanman replied:
ah, is this because when the app restarts it won't know the randomly generated secret? Hmm.. maybe we could store the secret somehow. If needed

colinrotherham replied:
Yeah that's the trade-off sadly.

The old approach re-hashed the secret on every restart. This would lock you out of your cookie causing it to be regenerated.

Sounds like an idea though.
